src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs

```diff
@@ -326,108 +310,74 @@ private BoundExpression BindSwitchExpression(SwitchExpressionSyntax node, Bindin
-                                Debug.Assert(!boundIndexerAccess.Indexer.IsStatic);
-                                ReportDiagnosticsIfObsolete(bindingDiagnostics, lengthProperty, syntax, hasBaseReceiver: false);
-                                GetWellKnownTypeMember(argIsIndex ? WellKnownMember.System_Index__ctor : WellKnownMember.System_Range__ctor, bindingDiagnostics, syntax: syntax);
-                                indexerAccess = BindIndexerDefaultArguments(boundIndexerAccess, BindValueKind.RValue, bindingDiagnostics);
-                                found = true;
-                                break;
-                            }
-                            found = false;
-                            break;
+            var diagnostics = BindingDiagnosticBag.GetInstance();
+            var success = BindLengthAndIndexerForListPattern(node, inputType, inputValEscape: ExternalScope, diagnostics, indexerAccess: out _, out var lengthAccess, receiverPlaceholder: out _, argumentPlaceholder: out _);
+            lengthProperty = success ? GetPropertySymbol(lengthAccess, out _, out _) : null;
+            diagnostics.Free();
+            return success;
+        }
 
-                        case BoundIndexOrRangePatternIndexerAccess boundIndexOrRangePatternIndexerAccess:
-                            lengthProperty = boundIndexOrRangePatternIndexerAccess.LengthOrCountProperty;
-                            patternSymbol = boundIndexOrRangePatternIndexerAccess.PatternSymbol;
-                            found = true;
-                            break;
+        private bool BindLengthAndIndexerForListPattern(SyntaxNode node, TypeSymbol inputType, uint inputValEscape, BindingDiagnosticBag diagnostics,
+            out BoundExpression? indexerAccess, out BoundExpression? lengthAccess, out BoundListPatternReceiverPlaceholder? receiverPlaceholder, out BoundListPatternUnloweredIndexPlaceholder? argumentPlaceholder)
+        {
+            var bindingDiagnostics = BindingDiagnosticBag.GetInstance(diagnostics);
 
-                        case var v:
-                            throw ExceptionUtilities.UnexpectedValue(v.Kind);
-                    }
-                    analyzedArguments.Free();
-                    indexerGroup.Free();
-                    goto done;
+            receiverPlaceholder = new BoundListPatternReceiverPlaceholder(node, GetValEscape(inputType, inputValEscape), inputType) { WasCompilerGenerated = true };
+            bool hasErrors = false;
+            if (inputType.IsSZArray())
+            {
+                hasErrors |= !TryGetSpecialTypeMember(Compilation, SpecialMember.System_Array__Length, node, bindingDiagnostics, out PropertySymbol lengthProperty);
+                if (lengthProperty is not null)
+                {
+                    lengthAccess = new BoundPropertyAccess(node, receiverPlaceholder, lengthProperty, LookupResultKind.Viable, lengthProperty.Type);
```

Comment:
>lengthAccess [](http://example.com/codeflow?start=20&length=12)

It feels like this should be marked as compiler generated

Reply:
What's the advantage? The `lengthAccess` we get back from `TryBindLengthOrCount` below (line 352) will not be marked as compiler-generated.

Reply:
> What's the advantage? The lengthAccess we get back from TryBindLengthOrCount below (line 352) will not be marked as compiler-generated.

We should mark that one too.

Reply:
In general, all bound nodes that do not reflect the original syntax should be marked as compiler generated. This way SemanticModel can skip them and avoids confusing itself.